File: server/engine/fuzzy_match.py
"""Direct port of src/lib/fuzzyMatch.ts -- Dice-coefficient + Levenshtein header similarity."""

import re

_STOPWORDS = {"the", "a", "of", "name"}


def normalize_header(raw: str) -> str:
    text = raw.lower().strip()
    text = re.sub(r"[^a-z0-9\s]", " ", text)
    tokens = [t for t in re.split(r"\s+", text) if t and t not in _STOPWORDS]
    return " ".join(tokens)


def _tokenize(s: str) -> list[str]:
    return [t for t in re.split(r"\s+", s) if t]


def _dice_coefficient(a: str, b: str) -> float:
    if a == b:
        return 1.0
    if len(a) < 2 or len(b) < 2:
        return 1.0 if a == b else 0.0

    def bigrams(s: str) -> dict[str, int]:
        counts: dict[str, int] = {}
        for i in range(len(s) - 1):
            bg = s[i : i + 2]
            counts[bg] = counts.get(bg, 0) + 1
        return counts

    bigrams_a = bigrams(a)
    bigrams_b = bigrams(b)
    intersection = 0
    for bg, count_a in bigrams_a.items():
        count_b = bigrams_b.get(bg)
        if count_b:
            intersection += min(count_a, count_b)
    total_a = sum(bigrams_a.values())
    total_b = sum(bigrams_b.values())
    if total_a + total_b == 0:
        return 0.0
    return (2 * intersection) / (total_a + total_b)


def _levenshtein(a: str, b: str) -> int:
    m, n = len(a), len(b)
    if m == 0:
        return n
    if n == 0:
        return m
    dp = list(range(n + 1))
    for i in range(1, m + 1):
        prev = dp[0]
        dp[0] = i
        for j in range(1, n + 1):
            temp = dp[j]
            dp[j] = prev if a[i - 1] == b[j - 1] else 1 + min(prev, dp[j], dp[j - 1])
            prev = temp
    return dp[n]


def header_alias_score(header: str, alias: str) -> float:
    """Similarity score in [0, ~1.15] between a normalized header and a normalized alias."""
    max_len = max(len(header), len(alias)) or 1
    dice = _dice_coefficient(header, alias)
    lev = 1 - _levenshtein(header, alias) / max_len
    token_a = set(_tokenize(header))
    token_b = set(_tokenize(alias))
    contains_bonus = 0.15 if (alias in header or header in alias) else 0.0
    token_overlap_bonus = 0.1 if token_b and token_b.issubset(token_a) else 0.0
    return 0.6 * dice + 0.3 * lev + contains_bonus + token_overlap_bonus

File: src/lib/fuzzyMatch.ts
const STOPWORDS = new Set(['the', 'a', 'of', 'name']);

export function normalizeHeader(raw: string): string {
  return raw
    .toLowerCase()
    .trim()
    .replace(/[^a-z0-9\s]/g, ' ')
    .split(/\s+/)
    .filter((tok) => tok.length > 0 && !STOPWORDS.has(tok))
    .join(' ');
}

function tokenize(s: string): string[] {
  return s.split(/\s+/).filter(Boolean);
}

/** Dice's coefficient over bigrams of the token-joined string. */
function diceCoefficient(a: string, b: string): number {
  if (a === b) return 1;
  if (a.length < 2 || b.length < 2) return a === b ? 1 : 0;

  const bigrams = (s: string): Map<string, number> => {
    const map = new Map<string, number>();
    for (let i = 0; i < s.length - 1; i++) {
      const bg = s.slice(i, i + 2);
      map.set(bg, (map.get(bg) ?? 0) + 1);
    }
    return map;
  };

  const bigramsA = bigrams(a);
  const bigramsB = bigrams(b);
  let intersection = 0;
  for (const [bg, countA] of bigramsA) {
    const countB = bigramsB.get(bg);
    if (countB) intersection += Math.min(countA, countB);
  }
  const totalA = [...bigramsA.values()].reduce((s, v) => s + v, 0);
  const totalB = [...bigramsB.values()].reduce((s, v) => s + v, 0);
  if (totalA + totalB === 0) return 0;
  return (2 * intersection) / (totalA + totalB);
}

function levenshtein(a: string, b: string): number {
  const m = a.length;
  const n = b.length;
  if (m === 0) return n;
  if (n === 0) return m;
  const dp: number[] = new Array(n + 1);
  for (let j = 0; j <= n; j++) dp[j] = j;
  for (let i = 1; i <= m; i++) {
    let prev = dp[0];
    dp[0] = i;
    for (let j = 1; j <= n; j++) {
      const temp = dp[j];
      dp[j] = a[i - 1] === b[j - 1] ? prev : 1 + Math.min(prev, dp[j], dp[j - 1]);
      prev = temp;
    }
  }
  return dp[n];
}

/** Similarity score in [0, ~1.15] between a normalized header and a normalized alias. */
export function headerAliasScore(header: string, alias: string): number {
  const maxLen = Math.max(header.length, alias.length) || 1;
  const dice = diceCoefficient(header, alias);
  const lev = 1 - levenshtein(header, alias) / maxLen;
  const tokenA = new Set(tokenize(header));
  const tokenB = new Set(tokenize(alias));
  const containsBonus = header.includes(alias) || alias.includes(header) ? 0.15 : 0;
  const tokenOverlapBonus =
    tokenB.size > 0 && [...tokenB].every((t) => tokenA.has(t)) ? 0.1 : 0;
  return 0.6 * dice + 0.3 * lev + containsBonus + tokenOverlapBonus;
}
